src/theory/quantifiers/term_database.cpp

```diff
@@ -222,191 +222,312 @@ Node TermDb::getMatchOperator( Node n ) {
-                if( std::find( d_func_map_rel_dom[f][i].begin(), 
-                               d_func_map_rel_dom[f][i].end(), d_arg_reps[n][i] ) == d_func_map_rel_dom[f][i].end() ){
-                  d_func_map_rel_dom[f][i].push_back( d_arg_reps[n][i] );
-                }
+  d_op_nonred_count[f] = 0;
+  // get the matchable operators in the equivalence class of f
+  std::vector<TNode> ops;
+  ops.push_back(f);
+  if (options::ufHo())
+  {
+    ops.insert(ops.end(), d_ho_op_slaves[f].begin(), d_ho_op_slaves[f].end());
+  }
+  Trace("term-db-debug") << "computeUfTerms for " << f << std::endl;
+  unsigned congruentCount = 0;
+  unsigned nonCongruentCount = 0;
+  unsigned alreadyCongruentCount = 0;
+  unsigned relevantCount = 0;
+  eq::EqualityEngine* ee = d_quantEngine->getActiveEqualityEngine();
+  for (const Node& ff : ops)
+  {
+    std::map<Node, std::vector<Node> >::iterator it = d_op_map.find(ff);
+    if (it == d_op_map.end())
+    {
+      // no terms for this operator
+      continue;
+    }
+    Trace("term-db-debug") << "Adding terms for operator " << ff << std::endl;
+    for (const Node& n : it->second)
+    {
+      // to be added to term index, term must be relevant, and exist in EE
+      if (!hasTermCurrent(n) || !ee->hasTerm(n))
+      {
+        Trace("term-db-debug") << n << " is not relevant." << std::endl;
+        continue;
+      }
+
+      relevantCount++;
+      if (!isTermActive(n))
+      {
+        Trace("term-db-debug") << n << " is already redundant." << std::endl;
+        alreadyCongruentCount++;
+        continue;
+      }
+
+      computeArgReps(n);
+      Trace("term-db-debug") << "Adding term " << n << " with arg reps : ";
+      for (unsigned i = 0; i < d_arg_reps[n].size(); i++)
+      {
+        Trace("term-db-debug") << d_arg_reps[n][i] << " ";
+        if (std::find(d_func_map_rel_dom[f][i].begin(),
+                      d_func_map_rel_dom[f][i].end(),
+                      d_arg_reps[n][i])
+            == d_func_map_rel_dom[f][i].end())
+        {
+          d_func_map_rel_dom[f][i].push_back(d_arg_reps[n][i]);
+        }
+      }
+      Trace("term-db-debug") << std::endl;
+      Assert(ee->hasTerm(n));
+      Trace("term-db-debug") << "  and value : " << ee->getRepresentative(n)
+                             << std::endl;
+      Node at = d_func_map_trie[f].addOrGetTerm(n, d_arg_reps[n]);
+      Assert(ee->hasTerm(at));
+      Trace("term-db-debug2") << "...add term returned " << at << std::endl;
+      if (at != n && ee->areEqual(at, n))
+      {
+        setTermInactive(n);
+        Trace("term-db-debug") << n << " is redundant." << std::endl;
+        congruentCount++;
+        continue;
+      }
+      if (ee->areDisequal(at, n, false))
+      {
+        std::vector<Node> lits;
+        lits.push_back(NodeManager::currentNM()->mkNode(EQUAL, at, n));
+        bool success = true;
+        if (options::ufHo())
+        {
+          // operators might be disequal
+          if (ops.size() > 1)
```

Comment:
join ``if``

Reply:
I'd say that having the first condition as the HO option also does, but I don't feel strongly about it